File: assembler/bloom.py
"""Bloom: the mip chain that turns bright regions into a glow.

The algorithm is the standard dual-filter pyramid:

    downsample  full -> 1/2 -> 1/4 -> 1/8 -> 1/16 -> 1/32   (threshold at 1/2)
    upsample    1/32 -> 1/16 -> 1/8 -> 1/4 -> 1/2           (tent, additive)

and the result is left in mip 0 at half resolution, which the assembler samples
with linear filtering. Half res is not a compromise: bloom is a wide, soft
signal, and there is nothing at full resolution for it to represent.

Threshold is applied ONCE, on the way down into the first mip. Every pass after
that is blurring what the threshold already selected, so re-applying it would
just eat the glow it was meant to spread.

Everything here works in linear HDR. The reference blooms after tonemapping and
so has to inverse-tonemap on the way in and re-apply on the way out; assembling
in the sane order removes both round-trips and the precision they cost.

RESOURCES ARE LAZY. Nothing is allocated until the first process() call, so
bloom left switched off costs no VRAM at all -- which is why it can default to
off without the memory showing up anyway.
"""

from pathlib import Path

import moderngl

from shared.gl_utils import read_shader, tryset, quad_vbo, quad_vao

_SHADER_DIR = Path(__file__).parent / "shaders"
_SHARED_SHADER_DIR = Path(__file__).parent.parent / "shared" / "shaders"

#: Halvings below the source. Five reaches 1/32, which at any sane window size
#: is a handful of texels -- wide enough for the diffuse halo, and past the
#: point where another level would add anything visible.
MIP_LEVELS = 5

#: Matches the camera's targets: bloom reads linear HDR and writes linear HDR.
_HDR_DTYPE = 'f2'


class Bloom:
    def __init__(self, ctx):
        self.ctx = ctx
        self.downsample_program = None
        self.upsample_program = None
        self.quad_vbo = None
        self.downsample_vao = None
        self.upsample_vao = None

        #: Mip textures/FBOs, largest first. None until the first process().
        self._mip_textures = []
        self._mip_fbos = []
        #: Source size the chain was built for; a change reallocates.
        self._source_size = None

        # Shaders are NOT loaded here: the Assembler owns this object and
        # reloads it as part of its own reload(), so compiling in the
        # constructor would just do the same work twice at startup.

    # ------------------------------------------------------------------
    # Shader loading (isolated so hot-reload can re-run it)
    # ------------------------------------------------------------------

    def reload(self):
        """Reload both bloom shaders from disk. Safe to call mid-execution."""
        vert = str(_SHARED_SHADER_DIR / 'fullscreen_quad.vert')

        try:
            program = self.ctx.program(
                vertex_shader=read_shader(vert),
                fragment_shader=read_shader(str(_SHADER_DIR / 'bloom_downsample.frag')),
            )
            # Only replaced on success: a failed compile leaves the old program
            # running rather than dropping bloom entirely.
            self.downsample_program = program
            if self.quad_vbo is None:
                self.quad_vbo = quad_vbo(self.ctx)
            # The VAO binds a program, so it must be rebuilt with the new one.
            self.downsample_vao = quad_vao(self.ctx, program, self.quad_vbo)
            print("Bloom downsample shader reloaded successfully")
        except Exception as e:
            print(f"Failed to reload bloom downsample shader: {e}")

        try:
            program = self.ctx.program(
                vertex_shader=read_shader(vert),
                fragment_shader=read_shader(str(_SHADER_DIR / 'bloom_upsample.frag')),
            )
            self.upsample_program = program
            if self.quad_vbo is None:
                self.quad_vbo = quad_vbo(self.ctx)
            self.upsample_vao = quad_vao(self.ctx, program, self.quad_vbo)
            print("Bloom upsample shader reloaded successfully")
        except Exception as e:
            print(f"Failed to reload bloom upsample shader: {e}")

    # ------------------------------------------------------------------
    # The chain
    # ------------------------------------------------------------------

    def process(self, source_texture, threshold, radius):
        """Build the bloom for `source_texture`. Returns the half-res result.

        Returns None if the shaders failed to compile, which the assembler
        reads as "no bloom this frame" rather than as an error.
        """
        if self.downsample_vao is None or self.upsample_vao is None:
            return None

        width, height = source_texture.size
        if width <= 0 or height <= 0:
            return None
        self._ensure_resources(width, height)
        if not self._mip_textures:
            return None

        # -- down: source -> mip 0 -> mip 1 -> ... --
        # Straight into mip 0 from the source. The reference blits to a
        # full-res copy first purely to get a matching format; ours already
        # matches, so that pass is pure cost.
        for i in range(MIP_LEVELS):
            src = source_texture if i == 0 else self._mip_textures[i - 1]
            src_w, src_h = src.size

            self._mip_fbos[i].use()
            src.use(location=0)
            tryset(self.downsample_program, 'source_tex', 0)
            tryset(self.downsample_program, 'apply_threshold', i == 0)
            tryset(self.downsample_program, 'threshold', float(threshold))
            tryset(self.downsample_program, 'source_texel_size',
                   (1.0 / src_w, 1.0 / src_h))
            self.downsample_vao.render(moderngl.TRIANGLES)

        # -- up: mip N -> mip N-1 -> ... -> mip 0, adding as it goes --
        self.ctx.enable(moderngl.BLEND)
        self.ctx.blend_func = moderngl.ONE, moderngl.ONE
        for i in range(MIP_LEVELS - 1, 0, -1):
            src = self._mip_textures[i]
            src_w, src_h = src.size

            self._mip_fbos[i - 1].use()
            src.use(location=0)
            tryset(self.upsample_program, 'source_tex', 0)
            tryset(self.upsample_program, 'source_texel_size',
                   (1.0 / src_w, 1.0 / src_h))
            tryset(self.upsample_program, 'bloom_radius', float(radius))
            self.upsample_vao.render(moderngl.TRIANGLES)
        self.ctx.disable(moderngl.BLEND)

        return self._mip_textures[0]

    # ------------------------------------------------------------------
    # Resources
    # ------------------------------------------------------------------

    def _ensure_resources(self, width, height):
        if self._source_size == (width, height):
            return
        self.release()

        mip_w, mip_h = width, height
        for _ in range(MIP_LEVELS):
            # max(1, ...) so a very small or very thin window cannot ask for a
            # zero-sized texture partway down the chain.
            mip_w = max(1, mip_w // 2)
            mip_h = max(1, mip_h // 2)
            texture = self.ctx.texture((mip_w, mip_h), 4, dtype=_HDR_DTYPE)
            texture.filter = (moderngl.LINEAR, moderngl.LINEAR)
            # Clamped, not repeated: the tent filter reaches past the edge, and
            # repeat would wrap a bright edge's glow around to the far side.
            texture.repeat_x = False
            texture.repeat_y = False
            self._mip_textures.append(texture)
            self._mip_fbos.append(
                self.ctx.framebuffer(color_attachments=[texture]))

        self._source_size = (width, height)

    def release(self):
        """Free the mip chain. Called on resize and at shutdown."""
        for fbo in self._mip_fbos:
            fbo.release()
        for texture in self._mip_textures:
            texture.release()
        self._mip_fbos = []
        self._mip_textures = []
        self._source_size = None

File: web/src/assembler/bloom.ts
/**
 * Bloom: the mip chain that turns bright regions into a glow.
 * A port of `assembler/bloom.py`.
 *
 * The algorithm is the standard dual-filter pyramid:
 *
 *     downsample  full -> 1/2 -> 1/4 -> 1/8 -> 1/16 -> 1/32   (threshold at 1/2)
 *     upsample    1/32 -> 1/16 -> 1/8 -> 1/4 -> 1/2           (tent, additive)
 *
 * and the result is left in mip 0 at half resolution, which the assembler
 * samples with linear filtering. Half res is not a compromise: bloom is a wide,
 * soft signal, and there is nothing at full resolution for it to represent.
 *
 * Threshold is applied ONCE, on the way down into the first mip. Every pass
 * after that is blurring what the threshold already selected, so re-applying it
 * would just eat the glow it was meant to spread.
 *
 * Everything here works in linear HDR. The original Fluoddity blooms after
 * tonemapping and so has to inverse-tonemap on the way in and re-apply on the
 * way out; assembling in the sane order removes both round-trips and the
 * precision they cost.
 *
 * RESOURCES ARE LAZY. Nothing is allocated until the first `process()` call, so
 * bloom left switched off costs no VRAM at all -- which is why it can default
 * to off without the memory showing up anyway.
 *
 * ## What is pre-built, and what is not
 *
 * The nine passes read nine different textures, so the texture half of each
 * bind group cannot be shared. All nine are built ONCE per resize and reused;
 * the uniform slices ride one buffer each with a dynamic offset. Nothing is
 * allocated per frame.
 */

import { compileModule } from '../gpu/shaderModule.ts';
import { HDR_FORMAT } from '../app/renderTargets.ts';
import { alignTo } from '../particleSystem/uniforms.ts';
import { MIP_LEVELS, mipSizes } from './bloomChain.ts';
import {
  BLOOM_DOWNSAMPLE_UNIFORM_SIZE,
  BLOOM_UPSAMPLE_UNIFORM_SIZE,
  packBloomDownsampleUniforms,
  packBloomUpsampleUniforms,
} from './assemblerUniforms.ts';

import bloomDownsampleSource from './shaders/bloomDownsample.wgsl';
import bloomUpsampleSource from './shaders/bloomUpsample.wgsl';

interface Mip {
  readonly texture: GPUTexture;
  readonly view: GPUTextureView;
  readonly size: readonly [number, number];
}

export class Bloom {
  private readonly device: GPUDevice;
  private readonly sampler: GPUSampler;

  private downPipeline: GPURenderPipeline | null = null;
  private upPipeline: GPURenderPipeline | null = null;
  private downLayout: GPUBindGroupLayout | null = null;
  private upLayout: GPUBindGroupLayout | null = null;

  private readonly downUniforms: GPUBuffer;
  private readonly upUniforms: GPUBuffer;
  private readonly downStride: number;
  private readonly upStride: number;

  /** Mip textures, largest first. Empty until the first `process()`. */
  private mips: Mip[] = [];
  /** One per level: the group reading the level above (or the source). */
  private downGroups: GPUBindGroup[] = [];
  /** One per upsample step: the group reading the level below. */
  private upGroups: GPUBindGroup[] = [];
  /** Source size the chain was built for; a change reallocates. */
  private sourceSize: readonly [number, number] | null = null;
  /** The source view the groups were built against. */
  private sourceView: GPUTextureView | null = null;

  constructor(device: GPUDevice, sampler: GPUSampler) {
    this.device = device;
    // Shared with the rest of the HDR chain. Clamped, not repeated: the tent
    // filter reaches past the edge, and repeat would wrap a bright edge's glow
    // around to the far side (`bloom.py:160-163`).
    this.sampler = sampler;

    const align = device.limits.minUniformBufferOffsetAlignment;
    this.downStride = alignTo(BLOOM_DOWNSAMPLE_UNIFORM_SIZE, align);
    this.upStride = alignTo(BLOOM_UPSAMPLE_UNIFORM_SIZE, align);

    // Sized for the whole chain up front: they are a few kilobytes, and unlike
    // the mip textures their size does not depend on the window.
    this.downUniforms = device.createBuffer({
      label: 'bloom-downsample-uniforms',
      size: this.downStride * MIP_LEVELS,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
    });
    this.upUniforms = device.createBuffer({
      label: 'bloom-upsample-uniforms',
      size: this.upStride * MIP_LEVELS,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
    });
  }

  /**
   * Reload both bloom shaders. Safe mid-execution.
   *
   * Each half builds independently: a failure in one leaves the other's
   * pipeline alone rather than taking the whole chain down (`bloom.py:66-71`).
   */
  async reload(): Promise<void> {
    const device = this.device;
    const [downModule, upModule] = await Promise.all([
      compileModule(device, 'bloomDownsample.wgsl', bloomDownsampleSource),
      compileModule(device, 'bloomUpsample.wgsl', bloomUpsampleSource),
    ]);

    // Identical shape for both, so one description covers them.
    const layout = (label: string): GPUBindGroupLayout =>
      device.createBindGroupLayout({
        label,
        entries: [
          {
            binding: 0,
            visibility: GPUShaderStage.FRAGMENT,
            buffer: { type: 'uniform', hasDynamicOffset: true },
          },
          { binding: 1, visibility: GPUShaderStage.FRAGMENT, texture: {} },
          { binding: 2, visibility: GPUShaderStage.FRAGMENT, sampler: {} },
        ],
      });

    if (downModule !== null) {
      this.downLayout = layout('bloom-downsample');
      this.downPipeline = device.createRenderPipeline({
        label: 'bloom-downsample',
        layout: device.createPipelineLayout({ bindGroupLayouts: [this.downLayout] }),
        vertex: { module: downModule, entryPoint: 'fullscreen_vs' },
        fragment: {
          module: downModule,
          entryPoint: 'fs_main',
          // No blend: each downsample REPLACES its destination.
          targets: [{ format: HDR_FORMAT }],
        },
        primitive: { topology: 'triangle-strip' },
      });
    }

    if (upModule !== null) {
      this.upLayout = layout('bloom-upsample');
      this.upPipeline = device.createRenderPipeline({
        label: 'bloom-upsample',
        layout: device.createPipelineLayout({ bindGroupLayouts: [this.upLayout] }),
        vertex: { module: upModule, entryPoint: 'fullscreen_vs' },
        fragment: {
          module: upModule,
          entryPoint: 'fs_main',
          targets: [
            {
              format: HDR_FORMAT,
              // The blend unit does the addition -- see bloomUpsample.wgsl.
              blend: {
                color: { srcFactor: 'one', dstFactor: 'one', operation: 'add' },
                alpha: { srcFactor: 'one', dstFactor: 'one', operation: 'add' },
              },
            },
          ],
        },
        primitive: { topology: 'triangle-strip' },
      });
    }

    // The layout objects are new, so every group built against the old ones is
    // stale. Dropping the source size forces a full rebuild on next process().
    this.invalidate();
  }

  pipelineStatus(): Readonly<Record<string, boolean>> {
    return {
      bloomDownsample: this.downPipeline !== null,
      bloomUpsample: this.upPipeline !== null,
    };
  }

  /** Drop the bind groups. Call after a resize, or when the source view changes. */
  invalidate(): void {
    this.downGroups = [];
    this.upGroups = [];
    this.sourceView = null;
  }

  /**
   * Build the bloom for `source`. Returns the half-res result, or null.
   *
   * Null means "no bloom this frame" rather than an error -- the shaders failed
   * to compile, or the window has no area. `assembler.py:92-95` reads it the
   * same way, and invariant 5 is why: a compile failure is logged, not fatal.
   */
  process(
    encoder: GPUCommandEncoder,
    source: GPUTextureView,
    sourceSize: readonly [number, number],
    threshold: number,
    radius: number,
  ): GPUTextureView | null {
    if (this.downPipeline === null || this.upPipeline === null) return null;
    if (this.downLayout === null || this.upLayout === null) return null;
    const [width, height] = sourceSize;
    if (width <= 0 || height <= 0) return null;

    this.ensureResources(width, height);
    if (this.mips.length === 0) return null;
    this.ensureGroups(source);

    // Uniforms for all nine passes, written before any pass is recorded --
    // `queue.writeBuffer` cannot interleave with an open encoder, the same
    // constraint the engine's sub-step loop solves the same way.
    const downBytes = new Uint8Array(this.downStride * MIP_LEVELS);
    const upBytes = new Uint8Array(this.upStride * MIP_LEVELS);
    for (let i = 0; i < MIP_LEVELS; i++) {
      // The SOURCE's texel size: the level being read, not written.
      const src = i === 0 ? sourceSize : this.mips[i - 1]!.size;
      downBytes.set(
        new Uint8Array(
          packBloomDownsampleUniforms([1 / src[0], 1 / src[1]], threshold, i === 0),
        ),
        i * this.downStride,
      );
    }
    for (let i = MIP_LEVELS - 1; i > 0; i--) {
      const src = this.mips[i]!.size;
      upBytes.set(
        new Uint8Array(packBloomUpsampleUniforms([1 / src[0], 1 / src[1]], radius)),
        i * this.upStride,
      );
    }
    const queue = this.device.queue;
    queue.writeBuffer(this.downUniforms, 0, downBytes);
    queue.writeBuffer(this.upUniforms, 0, upBytes);

    // -- down: source -> mip 0 -> mip 1 -> ... --
    // Straight into mip 0 from the source. The original blits to a full-res
    // copy first purely to get a matching format; ours already matches, so that
    // pass is pure cost.
    for (let i = 0; i < MIP_LEVELS; i++) {
      const pass = encoder.beginRenderPass({
        label: `bloom-down-${i}`,
        colorAttachments: [
          {
            view: this.mips[i]!.view,
            // 'clear' is right here: each downsample fully replaces its target.
            clearValue: { r: 0, g: 0, b: 0, a: 1 },
            loadOp: 'clear',
            storeOp: 'store',
          },
        ],
      });
      pass.setPipeline(this.downPipeline);
      pass.setBindGroup(0, this.downGroups[i]!, [i * this.downStride]);
      pass.draw(4);
      pass.end();
    }

    // -- up: mip N -> mip N-1 -> ... -> mip 0, adding as it goes --
    for (let i = MIP_LEVELS - 1; i > 0; i--) {
      const pass = encoder.beginRenderPass({
        label: `bloom-up-${i}`,
        colorAttachments: [
          {
            view: this.mips[i - 1]!.view,
            // 'load', NOT 'clear'. The destination holds its own downsampled
            // content and this pass adds to it -- see bloomUpsample.wgsl.
            loadOp: 'load',
            storeOp: 'store',
          },
        ],
      });
      pass.setPipeline(this.upPipeline);
      pass.setBindGroup(0, this.upGroups[i]!, [i * this.upStride]);
      pass.draw(4);
      pass.end();
    }

    return this.mips[0]!.view;
  }

  private ensureResources(width: number, height: number): void {
    if (this.sourceSize !== null
      && this.sourceSize[0] === width
      && this.sourceSize[1] === height) {
      return;
    }
    this.release();

    for (const [w, h] of mipSizes(width, height)) {
      const texture = this.device.createTexture({
        label: `bloom-mip-${this.mips.length}`,
        size: { width: w, height: h },
        format: HDR_FORMAT,
        usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.TEXTURE_BINDING,
      });
      this.mips.push({ texture, view: texture.createView(), size: [w, h] });
    }
    this.sourceSize = [width, height];
  }

  private ensureGroups(source: GPUTextureView): void {
    if (this.sourceView === source && this.downGroups.length === MIP_LEVELS) return;

    const bind = (
      layout: GPUBindGroupLayout,
      label: string,
      uniforms: GPUBuffer,
      size: number,
      texture: GPUTextureView,
    ): GPUBindGroup =>
      this.device.createBindGroup({
        label,
        layout,
        entries: [
          { binding: 0, resource: { buffer: uniforms, size } },
          { binding: 1, resource: texture },
          { binding: 2, resource: this.sampler },
        ],
      });

    this.downGroups = [];
    for (let i = 0; i < MIP_LEVELS; i++) {
      this.downGroups.push(
        bind(
          this.downLayout!,
          `bloom-down-${i}`,
          this.downUniforms,
          BLOOM_DOWNSAMPLE_UNIFORM_SIZE,
          i === 0 ? source : this.mips[i - 1]!.view,
        ),
      );
    }

    // Index i reads mip i and writes mip i-1, so slot 0 is unused. Kept aligned
    // with the loop index rather than compacted, because an off-by-one between
    // "which mip do I read" and "which uniform slice do I bind" is exactly the
    // kind of error that produces a plausible-looking glow.
    this.upGroups = [];
    for (let i = 0; i < MIP_LEVELS; i++) {
      this.upGroups.push(
        bind(
          this.upLayout!,
          `bloom-up-${i}`,
          this.upUniforms,
          BLOOM_UPSAMPLE_UNIFORM_SIZE,
          this.mips[i]!.view,
        ),
      );
    }

    this.sourceView = source;
  }

  /** Free the mip chain. Called on resize and at shutdown. */
  release(): void {
    for (const mip of this.mips) mip.texture.destroy();
    this.mips = [];
    this.sourceSize = null;
    this.invalidate();
  }

  destroy(): void {
    this.release();
    this.downUniforms.destroy();
    this.upUniforms.destroy();
  }
}
